objects/src/notes/inputs.rs

```diff
@@ -0,0 +1,59 @@
+use super::{Digest, Felt, Hasher, NoteError, ZERO};
+
+/// Holds the inputs which are placed onto the stack before a note's script is executed.
+/// - inputs are stored in reverse order such that when they are pushed onto stack they are in
+///   the correct order
+/// - hash is computed from inputs in the order they are stored (reverse order)
+#[derive(Debug, Eq, PartialEq)]
+pub struct NoteInputs {
+    inputs: [Felt; 16],
+    hash: Digest,
+}
+
+impl NoteInputs {
+    /// Number of note inputs.
+    const NOTE_NUM_INPUTS: usize = 16;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns NoteInputs created from the provided inputs.
+    ///
+    /// The inputs are padded with ZERO such that they are always of length 16.
+    ///
+    /// # Errors
+    /// Returns an error if the number of provided inputs is greater than 16.
+    pub fn new(inputs: &[Felt]) -> Self {
+        if inputs.len() > Self::NOTE_NUM_INPUTS {
+            NoteError::too_many_inputs(inputs.len());
+        }
+
+        // pad inputs with ZERO to be constant size (16 elements)
+        let mut padded_inputs = [ZERO; Self::NOTE_NUM_INPUTS];
+
+        // insert inputs starting from the end of the array
+        for (i, input) in (0..Self::NOTE_NUM_INPUTS).rev().zip(inputs.iter()) {
+            padded_inputs[i] = *input;
+        }
```

Comment:
I'm still not sure this is correct. Could we add a test to check it? Basically, I think the input `(a, b)` should result in the following array:
```
[0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, a, b]
```
The reason for this arrangement is that when we read the data back onto the stack using `mem_loadw` operations, the stack would be arranged as follows:
```
[b, a, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0,  0, 0, 0, 0, ...]
```
I think the code we need would look something like this:
```Rust
for (a, &b) in padded_inputs.iter_mut().zip(inputs.iter().rev()) {
  *a = b;
}
padded_inputs.reverse();
```
But I didn't test the above.

Reply:
Ok I see.  I was working under the assumption that inputs would be provided in the constructor in stack order but I believe they are actually provided in reverse stack order.  This makes sense now and is consistent with other parts of the codebase (`StackInputs`) we also assume they are passed in reverse stack order.  Please let me know if this is the correct mental model. I've amended the code with my new understanding and added a test.